### PR TITLE
ITNFA-17: Added support for connecting to the national server

### PR DESCRIPTION
**Pull request by oskarhinc**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/M2SysBiometricsConstants.java [hunk 1]
```diff
@@ -5,13 +5,20 @@
...
-    public static final String M2SYS_SERVER_URL = "m2sys-biometrics.server.url";
+    public static final String M2SYS_CLOUD_SCANNER_URL = "m2sys-biometrics.server.url";
 
     public static final String M2SYS_LOCAL_SERVICE_URL = "m2sys-biometrics.local-service.url";
 
-    public static final String M2SYS_USER = "m2sys-biometrics.server.user";
+    public static final String M2SYS_NATIONAL_SERVICE_URL = "m2sys-biometrics.national-service.url";
 
-    public static final String M2SYS_PASSWORD = "m2sys-biometrics.server.password";
+    public static final String M2SYS_LOCAL_USERNAME = "m2sys-biometrics.server.user";
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/M2SysBiometricsConstants.java]: These are the CloudScanner credentials - if you are renaming, rename to M2SYS_CLOUD_SCANNER_PASSWORD
and M2SYS_CLOUD_SCANNER_USER

Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/M2SysBiometricsConstants.java]: Ok, M2SYS_CLOUD_SCANNER_USER or M2SYS_CLOUD_SCANNER_USERNAME?
I believe M2SYS_CLOUD_SCANNER_USERNAME is more intuitive.

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/BioServerClient.java [hunk 2]
```diff
@@ -1,109 +1,23 @@
...
-    public String isRegistered(String serviceUrl, String subjectId) {
-        IsRegistered isRegistered = new IsRegistered();
-        isRegistered.setID(subjectId);
-
-        IsRegisteredResponse response = (IsRegisteredResponse) getWebServiceTemplate()
-                .marshalSendAndReceive(serviceUrl, isRegistered);
-
-        return response.getIsRegisteredResult();
-    }
+    Token getToken(M2SysClient client);
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/BioServerClient.java]: The token is only for CloudScanr. It does not belong in the BioServerClient, which is used to talk with the BioPluginServer by SOAP.

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/client/AbstractM2SysClient.java [hunk 3]
```diff
@@ -106,25 +119,10 @@ protected int getLocationID() {
     }
 
     protected Token getToken() {
-        String username = getProperty(M2SysBiometricsConstants.M2SYS_USER);
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/client/AbstractM2SysClient.java:109]: This should stay in this class

File: omod/src/main/resources/config.xml [hunk 4]
```diff
@@ -70,23 +70,39 @@
...
+            The m2sys cloud scanner url.
         </description>
     </globalProperty>
 
     <globalProperty>
         <property>${project.parent.artifactId}.server.user</property>
         <defaultValue></defaultValue>
         <description>
-            The m2sys server username.
+            The m2sys local (local to the clinic) server username.
```
Comment (pgesek) [inline omod/src/main/resources/config.xml]: This is the CloudScanr username

File: omod/src/main/resources/config.xml [hunk 5]
```diff
@@ -70,23 +70,39 @@
...
+        </description>
+    </globalProperty>
+
+    <globalProperty>
+        <property>${project.parent.artifactId}.server.national-user</property>
```
Comment (pgesek) [inline omod/src/main/resources/config.xml]: lets use a dot after national (national.user)

File: omod/src/main/resources/config.xml [hunk 6]
```diff
@@ -70,23 +70,39 @@
...
         <property>${project.parent.artifactId}.server.password</property>
         <defaultValue></defaultValue>
         <description>
-            The m2sys server password.
+            The m2sys local (local to the clinic) server password.
+        </description>
+    </globalProperty>
+
+    <globalProperty>
+        <property>${project.parent.artifactId}.server.national-password</property>
```
Comment (pgesek) [inline omod/src/main/resources/config.xml]: lets use a dot after national (national.password)

File: omod/src/main/resources/config.xml [hunk 7]
```diff
@@ -70,23 +70,39 @@
...
         </description>
     </globalProperty>
 
     <globalProperty>
         <property>${project.parent.artifactId}.server.password</property>
         <defaultValue></defaultValue>
         <description>
-            The m2sys server password.
+            The m2sys local (local to the clinic) server password.
+        </description>
```
Comment (pgesek) [inline omod/src/main/resources/config.xml]: This is the CloudScanr password

**Review by pgesek — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/M2SysBiometricsConstants.java (hunk 1 above)
Comment (pgesek) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/M2SysBiometricsConstants.java]: username can stay

File: omod/src/main/resources/config.xml (hunk 5 above)
Comment (oskarhinc) [reply omod/src/main/resources/config.xml]: There will be a difference with local-user then, but if it's not a problem I can change it.

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/NationalBioServerClient.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    @Autowired
+    private M2SysHttpClient httpClient;
+
+    @Override
+    public String getServiceUrl(M2SysClient client) {
+        return client.getProperty(M2SysBiometricsConstants.M2SYS_NATIONAL_SERVICE_URL);
+    }
+
+    @Override
+    public Token getToken(M2SysClient client) {
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/NationalBioServerClient.java]: This is not the authentication that will take place for the national server. The national server is behind OpenHIM that requires BASIC authentication along the SOAP request. 

In other words, the calls to the webServiceTemplate need to be provide authentication information.

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/BioServerClient.java [hunk 9]
```diff
@@ -1,109 +1,20 @@
...
-    public String isRegistered(String serviceUrl, String subjectId) {
-        IsRegistered isRegistered = new IsRegistered();
-        isRegistered.setID(subjectId);
-
-        IsRegisteredResponse response = (IsRegisteredResponse) getWebServiceTemplate()
-                .marshalSendAndReceive(serviceUrl, isRegistered);
-
-        return response.getIsRegisteredResult();
-    }
+    String getServiceUrl(M2SysClient client);
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/BioServerClient.java]: This seems more like an internal method - it belongs more in the abstract class, not the interface

Comment (pgesek) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/BioServerClient.java]: Moreover the bioplugin server should not use the client, the client should use the server. If the implementations need the property, they should get it from the admin service.

The get property mehod is something that should be part of the client interface it should stay private.

**Review by pgesek — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/BioServerClient.java [hunk 10]
```diff
@@ -1,109 +1,18 @@
...
-    public String isRegistered(String serviceUrl, String subjectId) {
-        IsRegistered isRegistered = new IsRegistered();
-        isRegistered.setID(subjectId);
-
-        IsRegisteredResponse response = (IsRegisteredResponse) getWebServiceTemplate()
-                .marshalSendAndReceive(serviceUrl, isRegistered);
-
-        return response.getIsRegisteredResult();
-    }
+    String getServiceUrl();
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/BioServerClient.java]: as mentioned, should not be part of the interface

**Review by pgesek — APPROVED** [review]

**Pull request merged**